aten/src/ATen/native/quantized/cpu/qconv.cpp

```diff
@@ -244,6 +265,155 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+              .first
+        : std::numeric_limits<uint8_t>::min();
+    auto output_max = ReluFused
+        ? activationLimits(output_scale, output_zero_point, Activation::RELU)
+              .second
+        : std::numeric_limits<uint8_t>::max();
+    qnnpack::conv_param_t conv_p(
+        {kernel_w, kernel_h},
+        {stride_w, stride_h},
+        {dilation_w, dilation_h},
+        {pad_t, pad_l, pad_t, pad_l},
+        groups,
+        in_ch,
+        out_ch,
+        kernel_zp,
+        kernel_scale,
+        output_min,
+        output_max);
+
+    auto outShape =
+        convOutputShape(N, H, W, K, kernel, stride, padding, dilation);
+    TORCH_CHECK(
+        std::all_of(
+            outShape.begin(), outShape.end(), [](int64_t i) { return i > 0; }),
+        "quanitzed::conv2d (qnnpack): each dimension of output tensor should be greater "
+        "than 0")
+    TORCH_CHECK(
+        (outShape[3] == out_ch),
+        "quanitzed::conv2d (qnnpack): Number of filters must be equal to number of "
+        "output channels")
+
+    // Allocate output Tensor and a buffer for QNNPACK to use
+    Tensor output = at::_empty_affine_quantized(
+        outShape,
+        at::device(kCPU).dtype(kQUInt8),
+        output_scale,
+        output_zero_point);
+
+    const pytorch_qnnp_status runStatus = qnnpack::qnnpackConv(
+        conv_p,
+        packB->getPackedWeights(),
+        N,
+        H,
+        W,
+        input_contig.q_scale(),
+        input_contig.q_zero_point(),
+        (uint8_t*)input_contig.data_ptr<c10::quint8>(),
+        output.q_scale(),
+        output.q_zero_point(),
+        (uint8_t*)output.data_ptr<c10::quint8>(),
+        nullptr);
+
+    TORCH_INTERNAL_ASSERT(
+        runStatus == pytorch_qnnp_status_success,
+        "failed to run quanitzed::conv2d (qnnpack) operator");
+
+    return output;
+  }
+#endif
+#if defined(USE_FBGEMM) || defined(USE_PYTORCH_QNNPACK)
+  Tensor operator()(
+      Tensor act,
+      Tensor packed_weight,
+      torch::List<int64_t> stride,
+      torch::List<int64_t> padding,
+      torch::List<int64_t> dilation,
+      int64_t groups,
+      double output_scale,
+      int64_t output_zero_point) {
+    auto& ctx = at::globalContext();
+#ifdef USE_FBGEMM
+    if (ctx.preferredQuantizedEngine() == at::QEngine::FBGEMM) {
+      return fbgemm_conv(
+          act,
+          packed_weight,
+          stride,
+          padding,
+          dilation,
+          groups,
+          output_scale,
+          output_zero_point);
+    }
+#endif
+#ifdef USE_PYTORCH_QNNPACK
+    if (ctx.preferredQuantizedEngine() == at::QEngine::QNNPACK) {
+      return qnnpack_conv(
+          act,
+          packed_weight,
+          stride,
+          padding,
+          dilation,
+          groups,
+          output_scale,
+          output_zero_point);
+    }
+#endif
+    TORCH_INTERNAL_ASSERT(
+        "Didn't find engine for operation quantized::conv ",
+        toString(ctx.preferredQuantizedEngine()));
```

Comment:
would be nice to have a constant declared somewhere for which engines are built in and print it here

Reply:
Yes, we can add that once we get the basic building blocks done :)